Migrate NSURLProtectionSpace serialization from WKKeyedCoder

sheeparegreat opened the pull request:
#### d6df66d73ecc084b401ad4b6f3ce29f14277151a
<pre>
Migrate NSURLProtectionSpace serialization from WKKeyedCoder
<a href="https://bugs.webkit.org/show_bug.cgi?id=278484">https://bugs.webkit.org/show_bug.cgi?id=278484</a>
<a href="https://rdar.apple.com/134452219">rdar://134452219</a>

Reviewed by Alex Christensen and Sihui Liu.

* Source/WebCore/platform/network/ProtectionSpaceBase.cpp:
(WebCore::ProtectionSpaceBase::isPasswordBased const):
* Source/WebCore/platform/network/ProtectionSpaceBase.h:
* Source/WebCore/platform/network/ProtectionSpaceBase.serialization.in:
* Source/WebCore/platform/network/cocoa/ProtectionSpaceCocoa.mm:
(WebCore::scheme):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLProtectionSpace.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCString.h.
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLProtectionSpace.mm: Added.
(WebKit::CoreIPCNSURLProtectionSpace::CoreIPCNSURLProtectionSpace):
(WebKit::CoreIPCNSURLProtectionSpace::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLProtectionSpace.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCString.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/WebCoreSupport/SocketStreamHandleImplCFNet.cpp:
(WebCore::authenticationSchemeFromAuthenticationMethod):
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(wkNSURLProtectionSpace_isEqual):
(TEST(IPCSerialization, Basic)):

Canonical link: <a href="https://commits.webkit.org/283373@main">https://commits.webkit.org/283373@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/343bf58403cb5bf8ff3c330c1671c34666df5f35

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66092 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45465 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18711 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70124 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16702 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68210 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53264 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16983 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53048 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11629 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69159 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41943 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57219 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33680 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38614 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14596 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15578 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60509 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14943 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71826 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10047 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/14357 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60368 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10079 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57281 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60658 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/14579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8298 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1941 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41273 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42349 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43532 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42093 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLProtectionSpace.serialization.in:
```diff
@@ -20,7 +20,7 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-#if PLATFORM(COCOA)
+#if PLATFORM(COCOA) && !HAVE(WK_SECURE_CODING_NSURLPROTECTIONSPACE)
```
Can we write like this:
> #if PLATFORM(COCOA)
> #if HAVE(WK_SECURE_CODING_NSURLPROTECTIONSPACE)
> ...
> #else
> ...
> #endif // HAVE(WK_SECURE_CODING_NSURLPROTECTIONSPACE)
> #endif // PLATFORM(COCOA)

sheeparegreat replied:
The generator doesn't support the `else` style yet.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLProtectionSpace.h:
```diff
@@ -0,0 +1,67 @@
+    double port;
+    double type;
+    std::optional<WebKit::CoreIPCString> realm;
+    double scheme;
```
This should not be a double.  Probably a string.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLProtectionSpace.h:
```diff
@@ -0,0 +1,67 @@
+struct CoreIPCNSURLProtectionSpaceData {
+    WebKit::CoreIPCString host;
+    double port;
+    double type;
```
This should probably be a WebCore::ProtectionSpaceBase::AuthenticationScheme

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLProtectionSpace.h:
```diff
@@ -0,0 +1,67 @@
+
+struct CoreIPCNSURLProtectionSpaceData {
+    WebKit::CoreIPCString host;
+    double port;
```
This should be a uint16_t or maybe an optional one.

szewai commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -549,8 +549,10 @@ static BOOL wkDDActionContext_isEqual(WKDDActionContext *a, SEL, WKDDActionConte
 {
-    if (![a.host isEqual: b.host])
-        return false;
+    if (![a.host isEqual: b.host]) {
```
`if (a.host != b.host && ![a.host isEqual: b.host])`
Why don't we need this check for the other params like protocol below?

sheeparegreat replied:
host is not nullable (the other members are). I added `(a.host == nil && b.host == nil)` as a special case for this edge case so the test passes (host should not be null in valid usage)

szewai commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLProtectionSpace.serialization.in:
```diff
@@ -34,3 +34,24 @@
+    WebKit::CoreIPCNSURLProtectionSpaceData m_data
+}
+
+#endif // PLATFORM(COCOA)
```
`// PLATFORM(COCOA) && HAVE(WK_SECURE_CODING_NSURLPROTECTIONSPACE)`

sheeparegreat replied:
updated

szewai commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLProtectionSpace.mm:
```diff
@@ -0,0 +1,131 @@
+    id scheme = dict[@"scheme"];
+    if ([scheme isKindOfClass:NSNumber.class]) {
+        auto val = [scheme unsignedCharValue];
+        if (isValidEnum<WebCore::ProtectionSpaceBaseAuthenticationScheme>(val))
```
I wonder how we are sure every scheme number must be converted to a valid enum. 
For example, if the framework makes a change to define a new scheme type with value 200 and WebKit does not update `ProtectionSpaceBaseAuthenticationScheme` accordingly, then the value will be dropped?

achristensen07 replied:
Yes.  This is desired.  If they add a new scheme, they will coordinate with us.

sheeparegreat committed:
Migrate NSURLProtectionSpace serialization from WKKeyedCoder
https://bugs.webkit.org/show_bug.cgi?id=278484
rdar://134452219

Reviewed by Alex Christensen and Sihui Liu.

* Source/WebCore/platform/network/ProtectionSpaceBase.cpp:
(WebCore::ProtectionSpaceBase::isPasswordBased const):
* Source/WebCore/platform/network/ProtectionSpaceBase.h:
* Source/WebCore/platform/network/ProtectionSpaceBase.serialization.in:
* Source/WebCore/platform/network/cocoa/ProtectionSpaceCocoa.mm:
(WebCore::scheme):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLProtectionSpace.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCString.h.
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLProtectionSpace.mm: Added.
(WebKit::CoreIPCNSURLProtectionSpace::CoreIPCNSURLProtectionSpace):
(WebKit::CoreIPCNSURLProtectionSpace::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLProtectionSpace.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCString.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/WebCoreSupport/SocketStreamHandleImplCFNet.cpp:
(WebCore::authenticationSchemeFromAuthenticationMethod):
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(wkNSURLProtectionSpace_isEqual):
(TEST(IPCSerialization, Basic)):

Canonical link: https://commits.webkit.org/283373@main
